New Apis

D4ve-R opened the pull request:
## Added
- `ModelApi` highlevel model api, accessible in `ifcApi.modelApi`
- `GeomApi` highlevel geometry api, accessible in `ifcApi.geomApi`
## Changes:
- `CreateModel` does NOT write ifc headers anymore.
- multi-line writing,`WriteLine`  accepts `Type | Type[]` now and returns `number | number[]`

## Example
The following file was written with the new api, see ModelApi & GeomApi tests
<img width="548" alt="Bildschirmfoto 2023-06-04 um 18 29 05" src="https://github.com/IFCjs/web-ifc/assets/69651599/67ed5a45-57a8-4bbf-a7a1-4749398ea5de">

D4ve-R commented on tests/functional/Columns.spec.ts:
```diff
@@ -45,6 +45,7 @@ describe('Test the column example', () => {
             count++;
         })
 		expect(count).toEqual(36);
-		expect(ifcAPI.GetAndClearErrors(m2).size()).toEqual(1);
```
@beachtom i suspect the parsing error was that no IfcProject entity is in the model, can you confirm ?

beachtom commented:
Really useful work, thanks!

I have left some comments on technical details.

But my overarching concern is hardcoding serveral of the ifc types into the modelAPI - when we have these classes in the schema already. My concern is maintaining the parity when we support new IFC versions. i.e. what if some value is added to one of these. I am wondering if there is some way we can massage the schema generation to enable us to use these... happy to take your thoughts.

I am not too concerned about the geometries as these are more low level.

beachtom commented:
So I like the changes. Good spot on removing the IDs in the schema file to reduce size (I assume that is the reason).

My only thoughts now are: can we have something in the schema gen. Perhaps in schema_aliases (we could perhaps rename that file schema_config) that list objects used in the API

This could then generate for each of them a typescript alias i.e.

 type IfcOwnerHistory = IFC2X3.IfcOwnerHistory | IFC4.IfcOwnerHistory | IFC4X3.IfcOwnerHistory

D4ve-R commented:
Oops, my bad, i didn't intent on pushing the changes in package.json & the build_scripts. 
I will check on the docker container, should be possible!

I think a huge benefit of removing the ID is that users don't need to care about them anymore, when creating new objects (they really shouldn't in most use-cases).
As you might have seen, i also removed a lot of special chars from the code generation, resulting in 1MB less data with still good readability.

> type IfcOwnerHistory = IFC2X3.IfcOwnerHistory | IFC4.IfcOwnerHistory | IFC4X3.IfcOwnerHistory

that was on my mind too! helps a lot with writing schema generic code

beachtom commented:
Great - thanks for removing that. I think this will be great once we can merge it.

A few other things I noticed:
- Commented out grid placement in geom api?
- Commented out IFCSIUNIT in model API
-The columns test is failing I suspect ( but am not certain) the writing out there needs to be converted to use the new model API?
-Thus the columns are also broken in the viewer 
-Examples/usage/src/modification.ts also seems to now be broken 

I think most of these are caused by the fact the CreateModel no longer creates the header. I think this is quite a breaking change and may cause a few people lots of issues. Perahps we should add a boolean parameter here that is default to true to outputs the header - and the ModelAPI can set this to false.

Happy to discuss alternative approaches as well

D4ve-R commented:
I will mark this as a real pull request, as soon as i'm ready to merge.

The columns test should work now, but i haven't tested the viewer examples. When i reopen the generated columns model i get parsing errors. Are you sure that the code generates valid ifc ? Because if yes, that means there might be a bug somewhere.

No all writing is still valid, there is no change in the api.

I also fixed the parsing error for empty sets and changed some test case, because they weren't working as expected.

beachtom commented:
Sorry I mean to reply to this - I will keep an eye out for you to mark it - but if I don't respind just @ me

beachtom commented:
@D4ve-R  - I need some of the code from this to be merged to help fixing a bug - do you mind if I extract elements of this and get it merged and you can finish the rest?

D4ve-R commented:
@beachtom no problem!
